README.md

```diff
@@ -100,51 +148,140 @@ shell>  cat info.log
 └ @ Main REPL[34]:4
 ```
 
-## `FilteredLogger`
+## `ActiveFilteredLogger`
 
-The `FilteredLogger` exists to give more control over which messages should be logged.
+The `ActiveFilteredLogger` exists to give more control over which messages should be logged.
 It warps any logger, and before sending messages to the logger to log,
 checks them against a filter function.
 The filter function takes the full set of parameters of the message.
-(See it's docstring with `?FilteredLogger` for more details.)
+(See it's docstring with `?ActiveFilteredLogger` for more details.)
 
 ### Demo
 We want to filter to only log strings staring with `"Yo Dawg!"`.
 
 ```
-julia> function yodawg_filter(level, message, _module, group, id, file, line; kwargs...)
-		startswith(msg, "Yo Dawg!")
+julia> function yodawg_filter(log_args)
+	startswith(log_args.message, "Yo Dawg!")
 end
  yodawg_filter (generic function with 1 method)                                                                                     
 
-julia> filtered_logger = FilteredLogger(yodawg_filter, global_logger());
+julia> filtered_logger = ActiveFilteredLogger(yodawg_filter, global_logger());
 
 julia> with_logger(filtered_logger) do
-		@info "Boring message"
-		@warn "Yo Dawg! it is bad"
-		@info "Another boring message"
-		@info "Yo Dawg! it is all good"
-	end
+	@info "Boring message"
+	@warn "Yo Dawg! it is bad"
+	@info "Another boring message"
+	@info "Yo Dawg! it is all good"
+end
 ┌ Warning: Yo Dawg! it is bad
 └ @ Main REPL[28]:3
 [ Info: Yo Dawg! it is all good
 ```
 
+## `EarlyFilteredLogger`
 
+The `EarlyFilteredLogger` is similar to the `ActiveFilteredLogger`,
+but it runs earlier in the logging pipeline.
+In particular it runs before the message is computed.
+It is (theoretically) useful to filter things early if creating the log message is expective.
```

Comment:
I've now seen Memento used *a lot*.
It doesn't support this and I am yet to hear complaints
but fair, will change to

```suggestion
It can be useful to filter things early if creating the log message is expensive.
```

Reply:
I dunno; I've just seen this actually happen in other systems which was why I tried to "design it away" in the logging frontend. Unfortunately it introduces some complexity and personally I don't feel completely happy with the way `shouldlog` and `min_enabled_level` turned out.

To explain some of the reasoning behind these functions:
* The reason for `min_enabled_level` was to allow for some very basic filtering without invoking dynamic dispatch at all becuase the `min_enabled_level` is cached in the logstate.
* On the other hand, `shouldlog` must invoke dynamic dispatch because the logger is loaded from the task local storage and can never be inferred. This is annoying.

So there's a choice here: which parts of early filtering go into some inflexible statically defined rules (currently only a level for all modules), which parts go into dynamically chosen rules (`shouldlog`) and which parts must be deferred until after message creation (`handle_message`)? If there's a way to integrate these more neatly that would be great.